### PR TITLE
Fix numpydoc parameter list handling

### LINKED ISSUE #54: numpydoc: Return types without description not displayed correctly
I am using numpydoc style docstrings. But when undocumented return types are used, then the type is displayed as normal text. For example in:
```python
"""
Returns
-------
None
"""
```
When adding a description, then it is rendered correctly:

```python
"""
Returns
-------
None
    Nothing to return.
"""
```

I am using pdoc 0.5.4.

### PR DESCRIPTION
**Pull request by rkaminsk**
Fixes #54.

### CODE REVIEW

**Comment by rkaminsk** [comment]
Please wait before merging. I still have to fine-tune the regular expression for return lists.

**Comment by rkaminsk** [comment]
Done from my side. Please have a look.

File: pdoc/html_helpers.py [hunk 1]
```diff
@@ -114,58 +114,98 @@ def _deflist(name, type, desc,
...
+
+        ret = ""
+        if name:
+            ret += '**`{}`**'.format(name)
         if type:
-            return '**`{}`** :&ensp;{}\n:   {}\n\n'.format(name, type, desc)
-        return '**`{}`**\n:   {}\n\n'.format(name, desc)
+            ret += ' :&ensp;{}'.format(type) if ret else type
+        if desc:
+            ret += '\n:   {}\n'.format(desc) if ret else desc
```
Comment (kernc) [inline pdoc/html_helpers.py]: `desc` is always True. `desc` without either `name` or `type` should error?

Comment (rkaminsk) [reply pdoc/html_helpers.py]: I can add an assertion. The `if desc` part was just copy and paste.

File: pdoc/html_helpers.py [hunk 2]
```diff
@@ -114,58 +114,98 @@ def _deflist(name, type, desc,
...
-                  re.compile(r'^(\*{0,2}\w+(?:, \*{0,2}\w+)*)(?: ?: (.*)(?<!\.)$)?'
-                             r'((?:\n(?: {4}.*|$))*)', re.MULTILINE).sub,
-                  _numpy_params.__func__),
-              _seealso=partial(
-                  re.compile(r'(See Also\n-{8}\n)(?:((?:\n?[\w.]* ?: .*)+)|(.*))').sub,
-                  _numpy_seealso.__func__)):
+              _sections=partial(re.compile(
+                  r'^(Returns\n-{7}\n|Yields\n-{6}\n|Raises\n-{6}\n|Warns\n-{5}\n|'
+                  r'Parameters\n-{10}\n|Receives\n-{8}\n|Other Parameters\n-{16}\n|'
+                  r'See Also\n-{8}\n)'
```
[the remark this answers is not in the document]
Comment (rkaminsk) [reply pdoc/html_helpers.py]: I think that requiring 3 dashes should work fine.

Supporting arbitrary section titles does not work with the `_returns` regular expression. Too many lines would be interpreted as types without description. Numpydoc also has free form sections and those would get messed up. One could rewrite the `_sections` expression to require that the whole section looks like a definition list...

I think it is a good idea to compile the regular expressions only once. I would add them as class attributes (private function parameters are ugly). This also has the advantage that a user can simply change them if necessary for a project. I kept it like this to follow pdoc's coding style.

File: pdoc/html_helpers.py [hunk 3]
```diff
@@ -114,58 +114,98 @@ def _deflist(name, type, desc,
...
+                            re.compile(r'^(?P<name>\*{0,2}\w+(?:, \*{0,2}\w+)*)'
+                                       r'(?: ?: (?P<type>.*))?(?<!\.)$'
+                                       r'(?P<desc>(?:\n(?: {4}.*|$))*)', re.MULTILINE).sub,
+                            _numpy_params.__func__),
+                        # All kinds of numpydoc Returns
+                        # (similar to the above but lines without a colon are types here)
+                        _returns=partial(
+                            re.compile(r'^(?:(?P<name>\*{0,2}\w+(?:, \*{0,2}\w+)*)'
+                                       r'(?: ?: (?P<type>.*))|'
+                                       r'(?P<single_type>[^ ].*))'
```
[the remark this answers is not in the document]
Comment (rkaminsk) [reply pdoc/html_helpers.py]: Numpydoc also supports names without anything (in the attributes section). So we have to be careful here.

File: pdoc/html_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkaminsk) [reply pdoc/html_helpers.py]: I missed the Attributes and Members titles. These are "official" section titles.

**Comment by rkaminsk** [comment]
See below some code that shows potential problem and some possible workarounds.

```python
import re

r = re.compile(
    r'^(\w[ \w]*)\n-{3,}\s*\n'                   # start at header
    r'((?:^(?!\s*\w[ \w]*\n-{3,})'               # stop  at next header
    r'(?:\*{0,2}\w+(?:, \*{0,2}\w+)*(?: ?: .*)|' # match `name : type`
    r'[^ ].*)'                                   # match single `name` or `type`
    r'(?<!\.)$'                                  # discard lines ending in a `.`
    r'(?:\n(?: {4}.*|$))*'                       # match description
    r'\s*)+)'                                    # gobble whitespace
    r'^(?=\w[ \w]*\n-{3,}|\Z)',                  # stop  at next header
    re.MULTILINE)

s = """\
See Also
--------
See also sections have to be matched separately, now.

Parameters
----------
x: int
    desc x

y: int
    desc y

no desc

Returns
-------
out : some type
    desc
some type

Other
-----
As long as there is something that looks like sentence.
This section will not be matched.

Problem
-------
The expression will accept this as a definition list
because there is no period at the end of the line!
For example, discarding if there is any kind of punctuation at the end
or if there is more than one entry that does not match a variable name!

Yields
------
some type
    with
    a long
    description
"""

for m in r.finditer(s):
    print (m.group(1, 2))
```

**Comment by rkaminsk** [comment]
Of course one can also have an expression that handles the 
```
Returns
-------
return type
    optional desc
```
separately because it makes no sense to have more than one return type line. Which is probably best if we follow that route. In any case we will need different expressions for the different section types if we do not want to make the types dependent on the section titles.
PS: Sorry for the many messages.

**Comment by kernc** [comment]
I meant like this. It seems to work with your example if you make stated types match `[^\s]+`.

File: pdoc/html_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kernc) [reply pdoc/html_helpers.py]: > I think it is a good idea to compile the regular expressions only once. 

They are compiled only once. Python caches up to `re._MAXCACHE` compiled state machines, so calling simply `re.sub()` results in overhead of approximately one dictionary lookup.

>  This also has the advantage that a user can simply change them if necessary for a project.

Not recommended. None of this is public API. Too fragile. :sweat_smile:

**Comment by rkaminsk** [comment]
> I meant like this. It seems to work with your example if you make stated types match `[^\s]+`.

But it won't match something like this, which is in line with numpydoc:
```
Returns
-------
int or string
```

`[^\s]+` would work with what I am actually planning to use for my project:
```
Returns
-------
List[int]
    desc
```

**Comment by rkaminsk** [comment]
Maybe using another test?

```diff
diff --git a/pdoc/html_helpers.py b/pdoc/html_helpers.py
index 4dd143d..be79980 100644
--- a/pdoc/html_helpers.py
+++ b/pdoc/html_helpers.py
@@ -161,12 +161,21 @@ class _ToMarkdown:
         if section.title() == 'See Also':
             body = _seealso(body)
         else:
-            body = re.sub(r'^(\*{0,2}\w+(?:, \*{0,2}\w+)*)'
-                          r'(?: ?: (.*)(?<!\.)$)?(?<!\.)$'
-                          r'((?:\n(?: {4}.*|$))*)',
-                          partial(_ToMarkdown._numpy_params, section=section),
-                          body,
-                          flags=re.MULTILINE)
+            m = re.match(r'\A\s*^(?!^\*{0,2}\w+(?:, \*{0,2}\w+)* ?:)'
+                         r'()'                       # empty name
+                         r'(\w.*)(?<![.!?])$'        # type that does not look like a sentence
+                         r'((?:\n(?: {4}.*|$))*)\s*\Z', # description until end of section
+                         body,
+                         flags=re.MULTILINE)
+            if m:
+                body = _ToMarkdown._numpy_params(m, section)
+            else:
+                body = re.sub(r'^(\*{0,2}\w+(?:, \*{0,2}\w+)*)'
+                              r'(?: ?: (.*)(?<!\.)$)?(?<!\.)$'
+                              r'((?:\n(?: {4}.*|$))*)',
+                              partial(_ToMarkdown._numpy_params, section=section),
+                              body,
+                              flags=re.MULTILINE)
         return section + '\n-----\n' + body
·
     @staticmethod
```

**Comment by kernc** [comment]
Sorry for the late reply. :sweat_smile:

Your proposed addendum check wouldn't be of much help as it then matches also:
```
Shopping List
-------------
1. peas
    continuation line
2. lentils
```
Querying numpy, scipy, and sklearn projects as some of the largest adherents of numpydoc style shows `int or list of int`-like unnamed types aren't so common (with these projects, anyway).
```bash
$ grep -izoRIP '(Returns|Yields|Raises|Warns)\n----+\n[^:]+\n' numpy scipy scikit-learn

```
(No results.)

With PEP 484 type hints being preferred going forward, and with a workaround (namely, naming outputs) available, I don't feel too strongly about supporting this use case.

**Comment by rkaminsk** [comment]
It's up to you. It's easy to add a workaround for my use case because I modified the templates, anyway.

**Comment by kernc** [comment]
I added some more tests and went back with your two-regex approach as it at least seems more verbose. Hopefully, no severe regressions were introduced. :sweat_smile:  This should land in 0.6.2 in a few hours. Thanks!

**Comment by rkaminsk** [comment]
Okay, thanks for caring! :smile:
